api/v1beta1/awsloadbalancercontroller_types.go

```diff
@@ -0,0 +1,206 @@
+	AWSAddonWAFv1  AWSAddon = "AWSWAFv1"
+	AWSAddonWAFv2  AWSAddon = "AWSWAFv2"
+)
+
+// +kubebuilder:validation:Enum=Auto;Manual
+type SubnetTaggingPolicy string
+
+const (
+
+	// AutoSubnetTaggingPolicy enables automatic subnet tagging.
+	AutoSubnetTaggingPolicy SubnetTaggingPolicy = "Auto"
+
+	// ManualSubnetTaggingPolicy disables automatic subnet tagging.
+	ManualSubnetTaggingPolicy SubnetTaggingPolicy = "Manual"
+)
+
+// AWSLoadBalancerControllerSpec defines the desired state of AWSLoadBalancerController
+type AWSLoadBalancerControllerSpec struct {
+
+	// SubnetTagging describes how resource tagging will be done by the operator.
+	//
+	// When in "Auto", the operator will detect the subnets where the load balancers
+	// will be provisioned and have the required resource tags on them. Whereas when
+	// set to manual, this responsibility lies on the user.
+	//
+	// +kubebuilder:default:=Auto
+	// +kubebuilder:validation:Optional
+	// +optional
+	SubnetTagging SubnetTaggingPolicy `json:"subnetTagging,omitempty"`
+
+	// Default AWS Tags that will be applied to all AWS resources managed by this
+	// controller (default []).
+	//
+	// This value is required so that this controller can function as expected
+	// in parallel to openshift-router.
+	//
+	// +kubebuilder:default:={}
+	// +kubebuilder:validation:Optional
+	// +optional
+	AdditionalResourceTags map[string]string `json:"additionalResourceTags,omitempty"`
+
+	// IngressClass specifies the Ingress class which the controller will reconcile.
+	// This Ingress class will be created unless it already exists.
+	// The value will default to "alb".
+	//
+	// +kubebuilder:default:=alb
+	// +kubebuilder:validation:Optional
+	// +optional
+	IngressClass string `json:"ingressClass,omitempty"`
+
+	// Config specifies further customization options for the controller's deployment spec.
+	//
+	// +kubebuilder:validation:Optional
+	// +optional
+	Config *AWSLoadBalancerDeploymentConfig `json:"config,omitempty"`
+
+	// AWSAddon describes the AWS services that can be integrated with
+	// the AWS Load Balancer.
+	//
+	// +kubebuilder:validation:Optional
+	// +optional
+	EnabledAddons []AWSAddon `json:"enabledAddons,omitempty"` // indicates which AWS addons should be disabled.
+
+	// Credentials is a reference to a secret containing
+	// the AWS credentials to be used by the controller.
+	// The secret is required to be in the operator namespace.
+	// If this field is empty - the credentials will be
+	// requested using the Cloud Credentials API,
+	// see https://docs.openshift.com/container-platform/4.11/authentication/managing_cloud_provider_credentials/about-cloud-credential-operator.html.
+	//
+	// +kubebuilder:validation:Optional
+	// +optional
+	Credentials *SecretReference `json:"credentials,omitempty"`
+}
+
+type AWSLoadBalancerDeploymentConfig struct {
+
+	// +kubebuilder:default:=2
+	// +kubebuilder:validation:Optional
+	// +optional
+	Replicas int32 `json:"replicas,omitempty"`
```

Comment:
Well, indeed, we didn't think of HPA at the time of writing of the operator. So, it'll be fighting with HPA.
How hard the requirement of being "HPA friendly" is?

Reply:
I was thinking about this in terms of which fields should be customizeable and how that would impact an API here.  I don't object per se, but deciding where you limit is worthwhile.  For instance, if this is the only value you want to allow customization for, there's no reason to nest it since .spec.replicas is common if not canonical.

Reply:
I see, yes for the moment we agreed on a single customization value - replicas. I believe we considered moving it up in the API  but based on the experience with the other operators where it was quite hard to argue with the customer about certain customizations (mostly scheduling: tolerations, nodeSelector, etc) we thought that this field would make sense to be a dedicated struct.